File: src/main/java/tk/loryruta/hgame/scenario/animation/Trigger.java
package tk.loryruta.hgame.scenario.animation;

import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.Input;

import java.util.function.Supplier;

/**
 * It's used by the update cycle to check periodically if some condition has been verified.
 * If returns {@code true}, the event can be executed. Otherwise the event can wait.
 */
public interface Trigger extends Supplier<Boolean> {
    Trigger NONE = () -> true;
    Trigger ENTER_KEY = () -> Gdx.input.isKeyJustPressed(Input.Keys.ENTER);

    static Trigger and(Trigger... triggers) {
        return () -> {
            for (Trigger t : triggers) {
                if (!t.get()) {
                    return false;
                }
            }
            return true;
        };
    }

    static Trigger sleep(long duration) {
        long now = System.currentTimeMillis();
        return () -> System.currentTimeMillis() >= (now + duration);
    }
}

File: src/main/kotlin/xyz/upperlevel/hgame/world/sequence/Trigger.kt
package xyz.upperlevel.hgame.world.sequence

import com.badlogic.gdx.Gdx
import com.badlogic.gdx.Input

/**
 * It's used by the update cycle to check periodically if some condition has been verified.
 * If returns `true`, the event can be executed. Otherwise the event can wait.
 */
typealias Trigger = () -> Boolean

object Triggers {
    val NONE = { true }
    val ENTER_KEY = { Gdx.input.isKeyJustPressed(Input.Keys.ENTER)}

    fun sleep(duration: Long): Trigger {
        val now = System.currentTimeMillis()
        return { System.currentTimeMillis() >= now + duration }
    }
}
